[#12] 동네 인증 도메인, 유즈케이스 구현 및 테스트 코드 작성

Srltas opened the pull request:
## 📌 Summary
사용자의 위치를 기반으로 동네를 인증하고 등록하는 기능에 대한 도메인, 유즈케이스를 구현하고 테스트 코드를 만들었습니다.

## 📝 Description
- **동네 인증**
  - 사용자가 특정 동네 범위 내에 있는지 확인하고, 그 결과에 따라 해당 동네를 인증할 수 있는 유즈케이스 구현
- **유연한 거리 계산**
  - 다양한 거리 계산 방식에 대응할 수 있도록 `DistanceCalculator` 인터페이스를 정의하고, 이를 구현하는 방식으로 설계
  (`Geoapify API`를 이용할려고 했으나, 두 위치의 직선 거리 계산은 지원하지 않아 `Haversine` 공식으로 직접 구현)
- **객체의 책임 분리**
  - `Neighborhood`클래스는 사용자의 현재 위치가 동네 범위 안에 있는지 검증하고, `Location`클래스는 두 위치 간의 거리를 계산하는 책임을 가짐
- [헥사고날 아키텍처를 기반으로 프로젝트 패키지 구조 설계 ](https://resonant-stoat-349.notion.site/b92705241e4243ea95a77d73d9b64026?pvs=74)

## ✅ Checklist
- [X] 새로운 기능이나 수정된 기능에 대해 충분한 테스트를 작성했습니다.
- [X] 코딩 스타일 가이드를 준수했습니다.
- [X] 문서(주석, README 등)가 필요하다면 업데이트했습니다.

Srltas committed:
feat: 동네인증 도메인, 유즈케이스 구현

- 다양한 거리 계산 방법에 대응할 수 있도록 'DistanceCalculator' 인터페이스 도입
- 'Neighborhood'가 자신의 범위를 확인할 수 있도록 비즈니스 로직을 내부에 포함
- 동네가 없거나 범위 밖인 경우를 처리하기 위해 커스텀 예외 생성

Srltas committed:
feat: 동네인증 테스트 코드 추가

f-lab-bradley reviewed (commented):
해당 프로젝트에 lombok 이 등록 안되어 있어요.
rebase 를 lombok 이 포함 된 브랜치로 진행 한 뒤 Lombok 브렌치가 main 으로 머지 되면 그 때 다시 main 으로 리베이스 해 주세요

f-lab-bradley commented on src/main/java/com/srltas/runtogether/domain/model/neighborhood/Neighborhood.java:
```diff
@@ -0,0 +1,26 @@
+
+@EqualsAndHashCode
+@AllArgsConstructor
+public class Neighborhood {
```
해당 클래스의 역할이 불분명 하네요.
멤버 변수로 Location 을 갖고 있는데  isWithinBoundary 에서는 전혀 상관 없는 두 Location 을 받는 것도 그렇구요..

Srltas replied:
Location과 Neighborhood 클래스는 다음과 같은 역할을 부여했습니다.
 - Location: 위도, 경도와 같은 위치 정보를 저장하는 역할입니다. 자신과 다른 위치 간의 거리를 계산하는 책임을 가지고 있습니다.
 - Neighborhood: 동네 이름, 설정된 동네 범위 등의 동네 정보를 저장합니다. 특정 위치가 해당 동네의 범위 안에 있는지 검증하는 책임을 맡고 있습니다.

isWithinBoundary 메소드는 위치 정보와 계산 방법을 파라미터로 받아, 자신의 범위 내에 있는지 검증하는 메소드입니다. 이 메소드는 동네 범위 검증에만 집중하기 위해, 두 위치 간의 거리 계산은 Location 클래스에게 위임하고, 반환된 거리를 이용해 자신이 설정한 범위를 넘지 않는지 검증합니다.

f-lab-bradley commented on src/main/java/com/srltas/runtogether/domain/model/neighborhood/Neighborhood.java:
```diff
@@ -0,0 +1,26 @@
+import lombok.Getter;
+
+@EqualsAndHashCode
+@AllArgsConstructor
```
AllArgsConstructor vs RequiredArgsConstructor 
두 어노테이션의 차이점을 알아보시고 작성 하신 것 과 같은 immutable class 에서는 어떤 어노테이션이 더 잘 어울릴지 고민 해 보시면 좋겠어요.

Srltas replied:
`@RequiredArgsConstructor` 어노테이션을 사용하는 것이 더 좋을거 같습니다.

`@AllArgsConstructor`는 모든 필드를 초기화하는 생성자를 만드는 것으로 객체의 불변성을 강조하기 보다는, 단순히 모든 필드를 초기화할 수 있는 생성자를 만드는 느낌을 줄거 같습니다. 하지만, `@RequiredArgsConstructor`는 필수 필드만 초기화하도록 강제하여, 클래스가 불변 객체임이 자연스럽게 표현됩니다.

f-lab-bradley commented on src/test/java/com/srltas/runtogether/application/NeighborhoodVerificationServiceTest.java:
```diff
@@ -0,0 +1,82 @@
+	}
+
+	@Test
+	public void testVerifyAndRegisterNeighborhood_WithinBoundary() {
```
동일한 함수를 사용 할 때 묶는 방법이 있어요.
@Nested 라는 어노테이션이 있는데요, 이거 한 번 알아보시고 괜찮으면 적용 해 보세요.
굳이 적용 안해도 되구요 ㅎㅎ

Srltas replied:
논리적으로 그룹화하여 더 명확하게 테스트를 표현할 수 있을거 같습니다.
한번 적용해 보겠습니다👍

Srltas committed:
[#13] Lombok 라이브러리 추가 (#14)

Srltas reviewed (commented):
패키지 구조나 클래스들의 위치에 대해서 궁금한 것이 있습니다.

두 위치 간의 거리를 계산하는 인터페이스(DistanceCalculator)는 Location 도메인과 밀접하게 관련되어 있다고 생각하여 `domain.model.location` 패키지에 생성 했습니다. 해당 인터페이스에 대한 구현체는 언제든 구현 방법이 변경될 수 있고, 외부 API를 사용할 가능성도 있다고 생각해 현재는 `adapter.out` 패키지에 위치 시켰습니다. 

하지만 이 구현체는 Location 도메인 내에서만 사용되며, 애플리케이션에서 사용될 가능성은 낮다고 생각합니다. 이 때문에, 해당 구현체의 패키지를 `domain.service.location`으로 옮겨 도메인과 관련된 기능을 한 곳에 모아두는 것이 더 나을지 고민이 됩니다. 멘토님께서는 이 부분에 대해 어떻게 생각하시는지 궁금합니다.

Srltas committed:
refactor: RequiredArgsConstructor로 변경

 - Lombok의 AllArgsConstructor 어노테이션을 RequiredArgsConstructor로 변경

Srltas committed:
test: 동네 인증 유즈케이스 테스트 코드 수정

Nested 애노테이션을 이용해 테스트 목적에 따라 분리

f-lab-bradley commented on src/main/java/com/srltas/runtogether/application/NeighborhoodVerificationService.java:
```diff
@@ -0,0 +1,31 @@
+
+	private final NeighborhoodRepository neighborhoodRepository;
+	private final VerifiedNeighborhoodRepository verifiedNeighborhoodRepository;
+	private final DistanceCalculator distanceCalculator;
```
Usecase 가 거리를 구하는 공식을 알 필요가 있을까요??
그것 보다 해당 클래스가 근처 동네를 구하는데 거리를 통해 구한다는 사실을 알고 있을 필요 있을까요??
이건 Neighborhood 클래스가 갖고 있게 하는건 어떻게 생각 하세요?

Srltas replied:
Neighborhood 클래스에서 거리를 구하는 공식을 알고 있는 것이 구현 세부 사항을 숨기고 코드의 응집도를 더 높여줄거 같습니다👍

f-lab-bradley commented on src/main/java/com/srltas/runtogether/domain/model/neighborhood/VerifiedNeighborhoodRepository.java:
```diff
@@ -0,0 +1,7 @@
+import com.srltas.runtogether.domain.model.user.User;
+
+public interface VerifiedNeighborhoodRepository {
+	void saveVerifiedNeighborhood(User user, Neighborhood neighborhood);
```
이럴땐 하나의 Aggregate 에 두 개념을 포함시켜서 한 번에 저장 하는 방식도 있어요.
DDD 의 Aggregate 에 대해 한 번 알아보시겠어요?

Srltas replied:
Aggregate 개념을 한번 적용해 봤습니다.

VerifiedNeighborhood 도메인을 새로 생성하고, 이를 User와 하나의 Aggregate로 묶었습니다.
User를 루트 엔티티로 설정함에 따라, VerifiedNeighborhoodRepository를 제거하고, UserRepository를 통해 User Aggregate 전체를 관리하도록 만들었습니다.

f-lab-bradley replied:
넵 지금은 맞는거 같아도 다음엔 또 달라질 수 있으니 너무 고민 하지말고 일단은 이렇게 가시죠 ㅎㅎㅎㅎ

f-lab-bradley reviewed (commented):
전체적으로 모킹해서 테스트도 잘 작성 해 주셨네요.
쪼끔만 더 다듬어 보아요

f-lab-bradley commented:
> 패키지 구조나 클래스들의 위치에 대해서 궁금한 것이 있습니다.
> 
> 두 위치 간의 거리를 계산하는 인터페이스(DistanceCalculator)는 Location 도메인과 밀접하게 관련되어 있다고 생각하여 `domain.model.location` 패키지에 생성 했습니다. 해당 인터페이스에 대한 구현체는 언제든 구현 방법이 변경될 수 있고, 외부 API를 사용할 가능성도 있다고 생각해 현재는 `adapter.out` 패키지에 위치 시켰습니다.
> 
> 하지만 이 구현체는 Location 도메인 내에서만 사용되며, 애플리케이션에서 사용될 가능성은 낮다고 생각합니다. 이 때문에, 해당 구현체의 패키지를 `domain.service.location`으로 옮겨 도메인과 관련된 기능을 한 곳에 모아두는 것이 더 나을지 고민이 됩니다. 멘토님께서는 이 부분에 대해 어떻게 생각하시는지 궁금합니다.

이거 답변 늦었죠;; commit 만 보다가 놓쳤어요... ㅈㅅㅈㅅ

일단 저는 domain 에 있어야 한다고 생각해요. 도메인에서도 외부 API 를 사용 할 수 있어요. 
다만 도메인에 설정된 인터페이스를 통해서 호출 하는 거구요...

도메인은 도메인에 선언된 interface 를 호출 하는 거고
adapter.out 은 해당 interface 의 구현체를 모아두는 것 이구요

Srltas committed:
refactor: 거리 계산을 Neighborhood 내부로 캡슐화

- 캡슐화와 코드의 응집도를 높이기 위한 수정입니다
- Neighborhood를 생성할 때 distanceCalculator를 주입 받습니다
- NeighborhoodVerificationService 유즈케이스에서 더 이상 distanceCalculator에 대해 알 필요가 없습니다

Srltas committed:
refactor: Aggregate 패턴 적용을 통한 코드 개선

- 인증된 동네를 관리하기 위해 VerifiedNeighborhood 도메인 생성
- User와 VerifiedNeighborhood를 하나의 애그리게이트로 통합
- User 도메인 내에 verifiedNeighborhood 컬렉션을 추가하여 인증된 동네 관리

f-lab-bradley reviewed (approved):
PR 에서 의도한 내용들 구현 완료 된건가요???
아니라면 드래프트로 바꿔주시고 맞으면 머지 진행 해 주세요.

Srltas commented:
> PR 에서 의도한 내용들 구현 완료 된건가요??? 아니라면 드래프트로 바꿔주시고 맞으면 머지 진행 해 주세요.

처음 목표 했던 동네 인증 도메인 모델, 유즈케이스, 테스트 케이스는 어느정도 구현이 되었다고 생가하여 우선 머지를 진행하도록 하겠습니다.
